app/javascript/packs/followButtons.js

```diff
@@ -0,0 +1,366 @@
+    formData.append('verb', verb);
+    getCsrfToken().then(sendFetch('follow-creation', formData));
+  }
+}
+
+/**
+ * Adds an event listener to the inner page content, to handle any and all follow button clicks with a single handler
+ */
+function listenForFollowButtonClicks() {
+  document
+    .getElementById('page-content-inner')
+    .addEventListener('click', handleFollowButtonClick);
+
+  document.getElementById(
+    'page-content-inner',
+  ).dataset.followClicksInitialized = true;
+}
+
+/**
+ * Sets the UI of the button based on the current following status
+ *
+ * @param {string} followStatus The current following status for the button
+ * @param {HTMLElement} button The button to update
+ */
+function updateInitialButtonUI(followStatus, button) {
+  const buttonInfo = JSON.parse(button.dataset.info);
+  const { style } = buttonInfo;
+  button.classList.add('showing');
+
+  switch (followStatus) {
+    case 'true':
+    case 'mutual':
+      updateFollowingButton(button, style);
+      break;
+    case 'follow-back':
+      addButtonFollowText(button, followStatus);
+      break;
+    case 'false':
+      updateFollowButton(button, 'follow', buttonInfo);
+      break;
+    case 'self':
+      updateUserOwnFollowButton(button);
+      break;
+    default:
+      addButtonFollowText(button, style);
+  }
+}
+
+/**
+ * Fetches all user 'follow statuses' for the given userIds, and then updates the UI for all buttons related to each user
+ *
+ * @param {Object} idButtonHash A hash of user IDs and the array buttons which relate to them
+ */
+function fetchUserFollowStatuses(idButtonHash) {
+  const url = new URL('/follows/bulk_show', document.location);
+  const searchParams = new URLSearchParams();
+  Object.keys(idButtonHash).forEach((id) => {
+    searchParams.append('ids[]', id);
+  });
+  searchParams.append('followable_type', 'User');
+  url.search = searchParams;
+
+  fetch(url, {
+    method: 'GET',
+    headers: {
+      Accept: 'application/json',
+      'X-CSRF-Token': window.csrfToken,
+      'Content-Type': 'application/json',
+    },
+    credentials: 'same-origin',
+  })
+    .then((response) => response.json())
+    .then((idStatuses) => {
+      Object.keys(idStatuses).forEach((id) => {
+        idButtonHash[id].forEach((button) => {
+          updateInitialButtonUI(idStatuses[id], button);
+        });
+      });
+    });
+}
+
+/**
+ * Sets up the initial state of all user follow buttons on the page,
+ * by obtaining the 'follow status' of each user and updating the associated buttons' UI.
+ */
+function initializeAllUserFollowButtons() {
+  const buttons = document.querySelectorAll(
+    '.follow-action-button.follow-user:not([data-fetched])',
+  );
+
+  if (buttons.length === 0) {
+    return;
+  }
+
+  const userIds = {};
+
+  Array.from(buttons, (button) => {
+    button.dataset.fetched = 'fetched';
+    const userStatus = document.body.getAttribute('data-user-status');
```

Comment:
Refactor (non-blocking): Prefer destructuring off the dataset property.

```suggestion
    const { userStatus } = document.body.dataset;
```